### PR TITLE
[FEATURE] Comment context

### PR DESCRIPTION
**Pull request by amalieboegild**
### Types of changes

- [ ] Fix (non-breaking change that fixes existing issue)
- [x] Feature (non-breaking change that adds a feature)
- [ ] Breaking change (fix or feature that breaks or removes other functionality)

### Pre-flight checklist

- [x] PR changes requires documentation change
  - [x] I have updated the documentation accordingly
- [x] I have tested my change, and run all unit tests 
- [x] I ran ``dotnet format`` and committed+pushed the result

## Description

The whole database layer for Comment runs now. Namely, Repository and Context. The CommentManager and ICommentService are also fully implemented. 
Also as agreed with Asger, there has been some changes to the Content system. 
Furthermore, this PR contains some API related changes. 

To sum up, the Comment subsystem is done now. Only small additions are needed, eg. shared Context for Comment and Content and avoid code duplication when setting up tests for CommentRepository and CommentManager. 

### Fixes issues:

- [x] #13 ]

## How has this been tested

All methods have been tested in the CommentRepositoryTests class.

## Screenshots of the solution (if appropriate)

### CODE REVIEW

File: SELearning.API/Controllers/CommentController.cs [hunk 1]
```diff
@@ -11,48 +12,128 @@ namespace SELearning.API.Controllers;
...
     [ProducesResponseType(404)] // Not Found
-    public async Task<ActionResult<CommentDTO>> GetComment(int id)
-        => (await _repository.GetAsync(id)).ToActionResult();
+    public async Task<ActionResult<Comment>> GetComment(int id)
+    {
+        try
+        {
+            return Ok(await _service.GetCommentFromCommentId(id));
+        }
+        catch (Exception)
```
Comment (albe2669) [inline SELearning.API/Controllers/CommentController.cs:33]: What happens here in case of an internal server error?

File: SELearning.Infrastructure/CommentManager.cs [hunk 2]
```diff
@@ -1,38 +1,87 @@
...
+                throw new Exception("The content that the comment belongs to could not be found");
+            }
         }
-        public void UpdateComment(Comment cmt)
+
+        public async Task UpdateComment(int id, CommentUpdateDTO dto)
         {
+            if ((await _repo.UpdateComment(id, dto)).Item1 == OperationResult.NotFound)
+            {
+                throw new Exception("The comment could not be found");
```
Comment (albe2669) [inline SELearning.Infrastructure/CommentManager.cs:25]: Can we create a custom exception or something instead of constantly repeating this? Also to make catching this specific exception easier?

File: SELearning.Infrastructure/CommentRepository.cs [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+            _context.Comments.Add(comment);
+
+            await _context.SaveChangesAsync();
+
+            CommentDetailsDTO dto = new CommentDetailsDTO(comment.Author, comment.Text, comment.Id, comment.Timestamp, comment.Rating, comment.Content);
+
+            return (OperationResult.Created, dto);
+        }
+
+        public async Task<(OperationResult, CommentDetailsDTO?)> UpdateComment(int Id, CommentUpdateDTO cmt)
```
Comment (albe2669) [inline SELearning.Infrastructure/CommentRepository.cs:37]: We some funky stuff here with nullability

File: SELearning.Infrastructure/CommentRepository.cs [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+
+            if (c == null)
+            {
+                return (OperationResult.NotFound, null);
+            }
+
+            c.Text = cmt.Text;
+            c.Rating = cmt.Rating;
+
+            CommentDetailsDTO dto = new CommentDetailsDTO(c.Author, c.Text, c.Id, c.Timestamp, c.Rating, c.Content);
```
Comment (albe2669) [inline SELearning.Infrastructure/CommentRepository.cs:49]: We some funky stuff here with nullability

File: SELearning.Infrastructure/CommentRepository.cs [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+
+            return (OperationResult.Deleted);
+        }
+
+        public async Task<Option<Comment>> GetCommentByCommentId(int commentId)
+        {
+            return await _context.Comments.FirstOrDefaultAsync(x => x.Id == commentId);
+        }
+
+        public async Task<(List<Comment>?, OperationResult)> GetCommentsByContentId(int contentId)
```
Comment (albe2669) [inline SELearning.Infrastructure/CommentRepository.cs:77]: We some funky stuff here with nullability

File: SELearning.Infrastructure/CommentRepository.cs [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+        {
+            Content? content = await _context.Content.FirstOrDefaultAsync(c => c.Id == contentId);
+            if (content == null)
+            {
+                return (null, OperationResult.NotFound);
+            }
+
+            List<Comment> comments = await _context.Comments.Where(x => x.Content.Id == contentId).ToListAsync();
```
Comment (albe2669) [inline SELearning.Infrastructure/CommentRepository.cs:85]: We some funky stuff here with nullability

**Review by albe2669 — CHANGES_REQUESTED** [review]
Apart from the comments here, it is worth noting in the description that this PR contains a bunch of API related changes, as well as some changes to the content system.

File: SELearning.API/Controllers/CommentController.cs (hunk 1 above)
Comment (amalieboegild) [reply SELearning.API/Controllers/CommentController.cs:33]: @JoachimBorup Det er din kode

File: SELearning.Infrastructure/CommentManager.cs (hunk 2 above)
Comment (amalieboegild) [reply SELearning.Infrastructure/CommentManager.cs:25]: This is already on my todo for tomorrow. I have talked to @YoHype as well, such that we will use the same exceptions for content and comment.

**Review by Arneproductions — APPROVED** [review]

File: SELearning.Infrastructure.Tests/CommentManagerTests.cs [hunk 7]
```diff
@@ -8,148 +8,129 @@ namespace SELearning.Infrastructure.Tests
...
 
-            //TODO: how do i assert that it throws something and what do we want it to throw?
         }
 
         [Fact]
         public void Update_given_new_content_succeeds()
         {
-            //TODO: how to do this? 
+
         }
```
Comment (avborup) [inline SELearning.Infrastructure.Tests/CommentManagerTests.cs:72]: Is it GitHub that shows this incorrectly or is this test empty? And the same for a test below.

File: SELearning.Infrastructure.Tests/CommentRepositoryTests.cs [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+            var connection = new SqliteConnection("Filename=:memory:");
+            connection.Open();
+            var builder = new DbContextOptionsBuilder<CommentContext>();
+            builder.UseSqlite(connection);
+            _context = new CommentContext(builder.Options);
+            _context.Database.EnsureCreated();
+
+            _repository = new CommentRepository(_context);
+
+            section.Content.Add(content);
```
Comment (avborup) [inline SELearning.Infrastructure.Tests/CommentRepositoryTests.cs:52]: There is a nullable thing warning here.

**Review by avborup — CHANGES_REQUESTED** [review]
I agree with @albe2669 that we should create a custom exception class - simply using `Exception` doesn't allow us to distinguish between our own exceptions and other exceptions in a try/catch. The warnings should also be fixed (except the one in `ContentRepositoryTests` - you haven't touched that code).

File: SELearning.API/Controllers/CommentController.cs (hunk 1 above)
Comment (JoachimBorup) [reply SELearning.API/Controllers/CommentController.cs:33]: The only way for me to know whether the `IContentService` fails or not is catching a generic `Exception` thrown by the service.
If an internal server error is thrown as an exception, then it's unfortunately caught and "presented" with the Not Found error code, and we'll probably need to make some custom exceptions as discussed [here](https://github.com/BDSA-2021/BDSA21-Say-AAAAAA/pull/28#discussion_r762416765), or figure something else out. If it's presented as something else than an exception, we can talk about it at the next meeting.

File: SELearning.Infrastructure.Tests/CommentManagerTests.cs (hunk 7 above)
Comment (amalieboegild) [reply SELearning.Infrastructure.Tests/CommentManagerTests.cs:72]: They are empty right now. Doing these last tests is the new thing on my todo. I just had to implement the methods in order for Joachim and me to merge, so this has not been 100% test driven development.

File: SELearning.Infrastructure.Tests/CommentRepositoryTests.cs (hunk 8 above)
Comment (amalieboegild) [reply SELearning.Infrastructure.Tests/CommentRepositoryTests.cs:52]: This is because Asger has declared almost properties in Content nullable. I don't think this is the right solution, since eg. id, author, etc. can't be null.

Comment (amalieboegild) [reply SELearning.Infrastructure.Tests/CommentRepositoryTests.cs:52]: But i think Asger is the one to solve it, when he merges his code into master?

File: SELearning.API/Controllers/CommentController.cs [hunk 9]
```diff
@@ -67,10 +92,20 @@ public async Task<IActionResult> CreateComment(int contentID, CommentDTO comment
     /// <returns>A response type 204: No Content if the comment exists, otherwise response type 404: Not Found.</returns>
     [AuthorizePermission(Permission.EditAnyComment)] // TODO: Create the possibility to have an 'or' evaluation of rules in the permission attribute and policy provider. EditOwnComment missing
     [HttpPut("{ID}")]
```
Comment (albe2669) [inline SELearning.API/Controllers/CommentController.cs:94]: This is fine

File: SELearning.API/Controllers/CommentController.cs [hunk 10]
```diff
@@ -82,5 +117,59 @@ public async Task<IActionResult> UpdateComment(int ID, CommentDTO comment)
...
+        }
+    }
+
+    /// <summary>
+    /// <c>UpvoteComment</c> increases the rating of the comment with the given ID.
+    /// </summary>
+    /// <param name="ID">The ID of the comment.</param>
+    /// <returns>A response type 204: No Content if the comment exists, otherwise response type 404: Not Found.</returns>
+    [Authorize]
+    [HttpPut("{ID}")]
```
Comment (albe2669) [inline SELearning.API/Controllers/CommentController.cs:138]: @JoachimBorup This will resolve to the same endpoint as update, Should probably be {ID}/upvote

File: SELearning.API/Controllers/CommentController.cs [hunk 11]
```diff
@@ -82,5 +117,59 @@ public async Task<IActionResult> UpdateComment(int ID, CommentDTO comment)
...
+        }
+    }
+
+    /// <summary>
+    /// <c>DownvoteComment</c> decreases the rating of the comment with the given ID.
+    /// </summary>
+    /// <param name="ID">The ID of the comment.</param>
+    /// <returns>A response type 204: No Content if the comment exists, otherwise response type 404: Not Found.</returns>
+    [Authorize]
+    [HttpPut("{ID}")]
```
Comment (albe2669) [inline SELearning.API/Controllers/CommentController.cs:160]: @JoachimBorup This will resolve to the same endpoint as update, Should probably be {ID}/upvote

**Review by albe2669 — CHANGES_REQUESTED** [review]

**Comment by albe2669** [comment]
Dont merge this branch until we can check it works with dotnet run

**Review by albe2669 — APPROVED** [review]
